File: core/src/main/java/com/vzome/core/math/symmetry/Symmetry.java

//(c) Copyright 2005, Scott Vorthmann.  All rights reserved.

package com.vzome.core.math.symmetry;

import java.util.Iterator;
import java.util.Set;

import com.vzome.core.algebra.AlgebraicField;
import com.vzome.core.algebra.AlgebraicMatrix;
import com.vzome.core.algebra.AlgebraicVector;
import com.vzome.core.math.RealVector;

/**
 * @author Scott Vorthmann
 *
 */
public interface Symmetry extends Iterable<Direction>, Embedding
{
	public enum SpecialOrbit {
	    BLUE,    // used for orbit dots
	    RED,     // used for orbit dots and in RotationTool, for the default rotation axis
	    YELLOW,  // used for orbit dots
	    BLACK    // if defined, used for orbit finding in getAxis
    }

    int PLUS = Axis.PLUS, MINUS = Axis.MINUS;
	
	int NO_ROTATION = -1;
    
    String TETRAHEDRAL = "tetrahedral", PYRITOHEDRAL = "pyritohedral";
    
    int getChiralOrder();
    
    String getName();

    Axis getAxis( AlgebraicVector vector );
    
    Axis getAxis( AlgebraicVector vector, OrbitSet orbits );

    Axis getAxis( RealVector vector, Set<Direction> filter );
    
    int getMapping( int from, int to );

    Permutation getPermutation( int i );

    AlgebraicMatrix getMatrix( int i );

	int inverse(int orientation);
	
    String[] getDirectionNames();

    Iterator<Direction> getDirections();
    
	Direction getDirection( String name );
    
    AlgebraicField getField();

    OrbitSet getOrbitSet();

    /**
     * Generate a subgroup, by taking the closure of some collection of Permutations
     * @param perms an array of Permutations indices
     * @return an array of Permutations indices
     */
    int[] closure( int[] perms );
    
    int[] subgroup( String name );
    
    Direction createNewZoneOrbit( String name, int prototype, int rotatedPrototype, AlgebraicVector vector );

    String getDefaultStyle();

    public abstract int[] getIncidentOrientations( int orientation );

    public abstract Direction getSpecialOrbit( SpecialOrbit which );

	Axis getPreferredAxis();

	/**
	 * Get the transformation matrix that maps zone 7 to zone -7 (for example).
	 * If null, the matrix is implicitly a central inversion, negating vectors.
	 * @return {@link AlgebraicMatrix}
	 */
	public abstract AlgebraicMatrix getPrincipalReflection();
	
	/**
	 * Compute the orbit triangle dots for all existing orbits,
	 * and leave behind an OrbitDotLocator for new ones.
	 * The result is just a VEF string, for debugging.
	 * @return
	 */
	public String computeOrbitDots();

    public boolean reverseOrbitTriangle();
}

File: online/src/worker/legacy/ts/com/vzome/core/math/symmetry/Symmetry.ts
import { java, javaemul } from "../../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { AlgebraicField } from "../../algebra/AlgebraicField.js";
import { AlgebraicMatrix } from "../../algebra/AlgebraicMatrix.js";
import { AlgebraicVector } from "../../algebra/AlgebraicVector.js";
import { Axis } from "./Axis.js";
import { Direction } from "./Direction.js";
import { Embedding } from "./Embedding.js";
import { OrbitSet } from "./OrbitSet.js";
import { Permutation } from "./Permutation.js";
import { SpecialOrbit } from "./SpecialOrbit.js";

/**
 * @author Scott Vorthmann
 * @class
 */
export interface Symmetry extends Embedding {
    getChiralOrder(): number;

    getName(): string;

    getAxis(vector?: any, orbits?: any): Axis;

    getMapping(from: number, to: number): number;

    getPermutation(i: number): Permutation;

    getMatrix(i: number): AlgebraicMatrix;

    inverse(orientation: number): number;

    getDirectionNames(): string[];

    getDirection(name: string): Direction;

    getField(): AlgebraicField;

    getOrbitSet(): OrbitSet;

    /**
     * Generate a subgroup, by taking the closure of some collection of Permutations
     * @param {int[]} perms an array of Permutations indices
     * @return {int[]} an array of Permutations indices
     */
    closure(perms: number[]): number[];

    subgroup(name: string): number[];

    createNewZoneOrbit(name: string, prototype: number, rotatedPrototype: number, vector: AlgebraicVector): Direction;

    getIncidentOrientations(orientation: number): number[];

    getSpecialOrbit(which: SpecialOrbit): Direction;

    getPreferredAxis(): Axis;

    /**
     * Get the transformation matrix that maps zone 7 to zone -7 (for example).
     * If null, the matrix is implicitly a central inversion, negating vectors.
     * @return {AlgebraicMatrix} {@link AlgebraicMatrix}
     */
    getPrincipalReflection(): AlgebraicMatrix;

    getOrbitTriangle(): AlgebraicVector[];

    /**
     * Compute the orbit triangle dots for all existing orbits,
     * and leave behind an OrbitDotLocator for new ones.
     * The result is just a VEF string, for debugging.
     * @return
     * @return {string}
     */
    computeOrbitDots(): string;

    reverseOrbitTriangle(): boolean;

    getDirections(): java.lang.Iterable<Direction>;
}

export namespace Symmetry {

    export const PLUS: number = 0;

    export const MINUS: number = 1;

    export const NO_ROTATION: number = -1;

    export const TETRAHEDRAL: string = "tetrahedral";

    export const PYRITOHEDRAL: string = "pyritohedral";
}
